### PR TITLE
Fix #4177: Generate optimised applyOrElse implementation for partial function literals

### LINKED ISSUE #4177: Partial Function: Prevent double evaluation of pattern matchers and guards
The documentation of `PartialFunction::applyOrElse` says:
> For all partial function literals the compiler generates an `applyOrElse` implementation which avoids double evaluation of pattern matchers and guards.

However this is not currently the case in Dotty:
```scala
object Test {
  private[this] var count = 0

  def test(x: Int) = { count += 1; true }

  object Foo {
    def unapply(x: Int): Option[Int] = { count += 1; Some(x) }
  }

  def main(args: Array[String]): Unit = {
    val res = List(1, 2).collect { case x if test(x) => x }
    println(count) // prints 4 but should be 2

    count = 0
    val res2 = List(1, 2).collect { case Foo(x) => x }
    println(count) // prints 4 but should be 2
  }
}
```

### PR DESCRIPTION
**Pull request by allanrenucci**
Fix #4177.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -585,6 +585,10 @@ class Definitions {
 
   lazy val PartialFunctionType: TypeRef         = ctx.requiredClassRef("scala.PartialFunction")
   def PartialFunctionClass(implicit ctx: Context) = PartialFunctionType.symbol.asClass
+
+    lazy val PartialFunction_applyOrElseR = PartialFunctionClass.requiredMethodRef(nme.applyOrElse)
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: This has the wrong indent size?

File: compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala [hunk 2]
```diff
@@ -54,38 +54,70 @@ class ExpandSAMs extends MiniPhase {
     val Block(
           (applyDef @ DefDef(nme.ANON_FUN, Nil, List(List(param)), _, _)) :: Nil,
           Closure(_, _, tpt)) = tree
-    val applyRhs: Tree = applyDef.rhs
+    val applyRhs = applyDef.rhs
     val applyFn = applyDef.symbol.asTerm
 
-    val MethodTpe(paramNames, paramTypes, _) = applyFn.info
+  val MethodTpe(paramNames, paramTypes, _) = applyFn.info
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala]: indent issue here too?

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 1 above)
Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: This is the indentation scheme used across the file

**Comment by allanrenucci** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 3]
```diff
@@ -273,8 +273,11 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
         coord = fns.map(_.pos).reduceLeft(_ union _))
     val constr = ctx.newConstructor(cls, Synthetic, Nil, Nil).entered
     def forwarder(fn: TermSymbol, name: TermName) = {
-      val fwdMeth = fn.copy(cls, name, Synthetic | Method).entered.asTerm
-      DefDef(fwdMeth, prefss => ref(fn).appliedToArgss(prefss))
+      var flags = Synthetic | Method
+      val isOverride = parents.exists(_.member(name).hasAltWith(_.info == fn.info))
```
[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: See `SymDenotation#allOverriddenSymbols`

**Comment by liufengyun** [comment]
@allanrenucci can you try rebase? It seems this PR has some interaction with `isInstanceofCheck` in the case of partial functions.

**Comment by liufengyun** [comment]
Thanks a lot @allanrenucci , I'm investigating it, type inference somehow failed to infer `a <: Int`, given `Some[a] <: X` and `X <: Some[Int]`.

**Comment by allanrenucci** [comment]
Blocked by #4297

File: compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala [hunk 4]
```diff
@@ -54,38 +54,72 @@ class ExpandSAMs extends MiniPhase {
...
+          ref(applyFn).appliedTo(paramRef)
+        }
+    }
+
     val isDefinedAtDef = transformFollowingDeep(DefDef(isDefinedAtFn, isDefinedAtRhs(_)))
-    val anonCls = AnonClass(tpt.tpe :: Nil, List(applyFn, isDefinedAtFn), List(nme.apply, nme.isDefinedAt))
-    cpy.Block(tree)(List(applyDef, isDefinedAtDef), anonCls)
+    val applyOrElseDef = transformFollowingDeep(DefDef(applyOrElseFn, applyOrElseRhs(_)))
+
+    val anonCls = AnonClass(tpt.tpe :: Nil, List(applyFn, isDefinedAtFn, applyOrElseFn), List(nme.apply, nme.isDefinedAt, nme.applyOrElse))
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala]: I don't think there's a reason. We could implement them directly.

**Comment by Blaisorblade** [comment]
> if the RHS of the anonymous function is a Match then it is not total otherwise it is

I assume you include `x => { x match { case 1 => x.toString } }` as a PartialFunction, per #4241.

-------

(Edited description so the PR is linked in #4177, "Fix #NNNN" in the title doesn't work).

**Pull request merged**